src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/DictionarySlim.cs

```diff
@@ -0,0 +1,544 @@
+            int entryIndex = GetEntryIndex(bucketIndex);
+
+            int lastIndex = -1;
+            while (entryIndex != -1)
+            {
+                if (entries[entryIndex].key.Equals(key))
+                {
+                    if (lastIndex != -1)
+                    {   // Fixup preceding element in chain to point to next (if any)
+                        entries[lastIndex].next = entries[entryIndex].next;
+                    }
+                    else
+                    {   // Fixup bucket to new head (if any)
+                        _buckets[bucketIndex] = entries[entryIndex].next + 1;
+                    }
+
+                    entries[entryIndex] = default; // could use RuntimeHelpers.IsReferenceOrContainsReferences
+
+                    entries[entryIndex].next = -3 - _freeList; // New head of free list
+                    _freeList = entryIndex;
+
+                    Count--;
+                    return true;
+                }
+                lastIndex = entryIndex;
+                entryIndex = entries[entryIndex].next;
+            }
+
+            return false;
+        }
+
+        public ref TValue this[TKey key]
+        {
+            get
+            {
+                Entry[] entries = _entries;
+                int bucketIndex = (int)(GetHashCode(key) % (uint)entries.Length);
+                int collisionCount = 0;
+                for (int i = GetEntryIndex(bucketIndex);
+                        (uint)i < (uint)entries.Length; i = entries[i].next)
+                {
+                    if (key.Equals(entries[i].key))
+                        return ref entries[i].value;
+                    if (collisionCount >= entries.Length)
+                    {
+                        // The chain of entries forms a loop; which means a concurrent update has happened.
+                        // Break out of the loop and throw, rather than looping forever.
+                        throw new InvalidOperationException(Strings.InvalidOperation_ConcurrentOperationsNotSupported);
+                    }
+                    collisionCount++;
+                }
+
+                return ref AddKey(key, bucketIndex);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private ref TValue AddKey(TKey key, int bucketIndex)
+        {
+            Entry[] entries = _entries;
+            int entryIndex;
+            if (_freeList != -1)
+            {
+                entryIndex = _freeList;
+                _freeList = -3 - entries[_freeList].next;
+            }
+            else
+            {
+                if (Count == entries.Length || entries.Length == 1)
+                {
+                    entries = Resize();
+                    bucketIndex = (int)(GetHashCode(key) % (uint)entries.Length);
+                    // entry indexes were not changed by Resize
+                }
+                entryIndex = Count;
+            }
+
+            entries[entryIndex].key = key;
```

Comment:
I don't think I can, it could have been copied by value. And I didn't have good results earlier when I tried to extract a ref local - turns out reindexing into an array was much faster (or the JIT was extracting it itself)

Reply:
As already described in a previous post: JIT's CSE usually handles such cases well enough and caching in a `ref` local may actually result in worse performance.